### PR TITLE
[refactor] Add state_dict to loops

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR adds the templating logic for loops state. This would be used by Fault Tolerant Logic.

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/loops/fit_loop.py [hunk 1]
```diff
@@ -274,3 +282,11 @@ def _check_checkpoint_callback(self, should_update: bool, is_last: bool = False)
 
             for cb in callbacks:
                 cb.on_validation_end(self.trainer, model)
+
+    def state_dict(self) -> Dict:
+        if not self.is_connected:
+            raise MisconfigurationException("The Trainer should be connected to loop to retrieve the state_dict.")
+        return {self.epoch_loop.name: self.epoch_loop.state_dict()}
```
Comment (carmocca) [inline pytorch_lightning/loops/fit_loop.py:289]: What is the benefit of letting each loop define its saved name when the parent gets the state dict?

File: pytorch_lightning/loops/fit_loop.py [hunk 2]
```diff
@@ -50,7 +51,7 @@ def __init__(
         super().__init__()
         self.max_epochs = 1000 if (max_epochs is None and max_steps is None) else max_epochs
         self.min_epochs = 1 if (min_epochs is None and min_steps is None) else min_epochs
-        self.epoch_loop = TrainingEpochLoop(min_steps, max_steps)
+        self.epoch_loop: Optional[TrainingEpochLoop] = TrainingEpochLoop(min_steps, max_steps)
```
Comment (carmocca) [inline pytorch_lightning/loops/fit_loop.py:54]: Why is it optional now?

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 3]
```diff
@@ -47,8 +49,8 @@ def __init__(self, min_steps: int, max_steps: int):
         self.batches_seen: int = 0
         self.is_last_batch: Optional[bool] = None
 
-        self.batch_loop: Optional[TrainingBatchLoop] = None
-        self.val_loop: Optional[loops.EvaluationLoop] = None
+        self.batch_loop = TrainingBatchLoop()
+        self.val_loop = loops.EvaluationLoop()
```
Comment (carmocca) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:53]: I like this, but means we should remove the creation in `connect`

And `connect` should take an optional argument for the user to pass a custom loop

`val_loop: Optional[Loop] = None`

and then

`self.val_loop = val_loop or self.val_loop`

File: pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 4]
```diff
@@ -266,3 +269,9 @@ def on_evaluation_epoch_end(self) -> None:
         self.trainer.call_hook(hook_name)
         self.trainer.call_hook("on_epoch_end")
         self.trainer.logger_connector.on_epoch_end()
+
+    def state_dict(self) -> Dict:
+        return {}
+
+    def load_state_dict(self, state_dict: Dict) -> None:
+        raise NotImplementedError
```
Comment (carmocca) [inline pytorch_lightning/loops/dataloader/evaluation_loop.py:277]: Why does this need to raise an Error? If there's is nothing to `load` then it can just `pass`

File: pytorch_lightning/loops/base.py [hunk 5]
```diff
@@ -59,6 +64,10 @@ def skip(self) -> bool:
     def connect(self, trainer: 'pl.Trainer', *args: Any, **kwargs: Any) -> None:
         """Connects Loop with all the necessary things like connectors and accelerators."""
         # TODO(@justusschock): Make the trainer a weakref/proxy
+        if not isinstance(trainer, pl.Trainer):
+            raise MisconfigurationException(
+                f"Loop {self.__class__.__name__} should be connected to a :class:`~pytorch_lightning.Trainer` instance."
+            )
```
Comment (carmocca) [inline pytorch_lightning/loops/base.py:70]: When would this happen? Just to make sure nobody passes the wrong arg order?

File: pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 6]
```diff
@@ -29,6 +29,8 @@
 class EvaluationLoop(DataLoaderLoop):
     """Loops over all dataloaders for evaluation."""
 
+    name = "val_loop"
```
Comment (awaelchli) [inline pytorch_lightning/loops/dataloader/evaluation_loop.py:32]: why these names?

this is not the val loop, this is the evaluation loop for both testing and validation.

File: pytorch_lightning/loops/fit_loop.py (hunk 1 above)
Comment (awaelchli) [inline pytorch_lightning/loops/fit_loop.py:289]: why is the name of the attribute in this instance be determined by the class of the TrainingEpochLoop?
we can name self.my_loop anything we want. This way of state dict loading will not work for loop customization unfortunately. 

I suggest this:


`return {"epoch_loop": self.epoch_loop.state_dict()}`

This way, the class who is dumping it's state dict is responsible for knowing what it's attribute name are.

**Review by ananthsub — COMMENTED** [review]
do you have examples of what types of state the loop needs to save/load?

**Comment by carmocca** [comment]
> do you have examples of what types of state the loop needs to save/load?

Namely progress (dataclasses) and results (`ResultCollection`)

```python
trainer
    loops:
        fit_loop:
            epoch_loop:
                batch_loop:
                validation_loop:
                    ...
        validate_loop:
            ...
        test_loop:
            ...
------
and each loop has:
a_loop:
    progress
    results
```

File: pytorch_lightning/loops/fit_loop.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/loops/fit_loop.py:289]: It thought it could be useful for tracking dynamically the names.

File: pytorch_lightning/loops/dataloader/evaluation_loop.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py:277]: Done !

File: pytorch_lightning/loops/base.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/loops/base.py:70]: Yes.

File: pytorch_lightning/loops/dataloader/evaluation_loop.py (hunk 6 above)
Comment (tchaton) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py:32]: Removed !

File: pytorch_lightning/loops/fit_loop.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/loops/fit_loop.py:289]: Yes, sounds better.

Comment (tchaton) [reply pytorch_lightning/loops/fit_loop.py:289]: Removed !

File: pytorch_lightning/loops/fit_loop.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/loops/fit_loop.py:54]: Removed !

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:53]: Let's take care of this when actually getting into customisation.

File: pytorch_lightning/loops/base.py [hunk 7]
```diff
@@ -59,6 +64,10 @@ def skip(self) -> bool:
     def connect(self, trainer: 'pl.Trainer', *args: Any, **kwargs: Any) -> None:
         """Connects Loop with all the necessary things like connectors and accelerators."""
         # TODO(@justusschock): Make the trainer a weakref/proxy
+        if not isinstance(trainer, pl.Trainer):
+            raise MisconfigurationException(
+                f"Loop {self.__class__.__name__} should be connected to a :class:`~pytorch_lightning.Trainer` instance."
```
Comment (carmocca) [inline pytorch_lightning/loops/base.py:69]: This is not a rendered doc

```suggestion
                f"Loop {self.__class__.__name__} should be connected to a `Trainer`, found: {trainer}."
```

File: pytorch_lightning/loops/base.py [hunk 8]
```diff
@@ -46,6 +47,10 @@ def __init__(self) -> None:
         self.iteration_count: int = 0
         self.trainer: Optional['pl.Trainer'] = None
 
+    @property
+    def is_connected(self) -> bool:
+        return self.trainer is not None
```
Comment (carmocca) [inline pytorch_lightning/loops/base.py:52]: Do we need a property for this? The check is so simple I don't think it's worth it.

Also the name is confusing as it only checks that the trainer is present, but more things could be "connected" down the line

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 9]
```diff
@@ -674,3 +674,9 @@ def _truncated_bptt_steps(self) -> int:
         if lightning_module.truncated_bptt_steps > 0:
             return lightning_module.truncated_bptt_steps
         return self.trainer.truncated_bptt_steps or 0
+
+    def state_dict(self) -> Dict:
+        return {}
+
+    def load_state_dict(self, state_dict: Dict) -> None:
+        pass
```
Comment (carmocca) [inline pytorch_lightning/loops/batch/training_batch_loop.py:682]: This would be inherited from the parent

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 10]
```diff
@@ -674,3 +674,9 @@ def _truncated_bptt_steps(self) -> int:
         if lightning_module.truncated_bptt_steps > 0:
             return lightning_module.truncated_bptt_steps
         return self.trainer.truncated_bptt_steps or 0
+
+    def state_dict(self) -> Dict:
+        return {}
```
Comment (carmocca) [inline pytorch_lightning/loops/batch/training_batch_loop.py:679]: The base class could return the empty dict itself

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:51]: We can wait for the `loop` argument but the creation in `connect` should be removed in this PR

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 11]
```diff
@@ -425,3 +425,10 @@ def _save_loggers_on_train_batch_end(self) -> None:
         should_flush_logs = self.trainer.logger_connector.should_flush_logs
         if should_flush_logs and self.trainer.is_global_zero and self.trainer.logger is not None:
             self.trainer.logger.save()
+
+    def state_dict(self) -> Dict:
+        return {"batch_loop": self.batch_loop.state_dict(), "validation_loop": self.val_loop.state_dict()}
```
Comment (carmocca) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:430]: Consistency

```suggestion
        return {"batch_loop": self.batch_loop.state_dict(), "val_loop": self.val_loop.state_dict()}
```

File: pytorch_lightning/loops/fit_loop.py [hunk 12]
```diff
@@ -97,6 +98,12 @@ def min_steps(self) -> int:
         """Returns the minimum numnber of steps to run"""
         return self.epoch_loop.min_steps
 
+    @min_steps.setter
+    def min_steps(self, value: int) -> None:
+        """Sets the minimum number of steps (forwards to epoch_loop)"""
+        # TODO(@awaelchli): This setter is required by debugging connector (fast dev run), should be avoided
+        self.epoch_loop.min_steps = value
+
```
Comment (carmocca) [inline pytorch_lightning/loops/fit_loop.py:106]: Why is this necessary now but not before?

File: pytorch_lightning/trainer/properties.py [hunk 13]
```diff
@@ -555,6 +555,13 @@ def _results(self) -> Optional[ResultCollection]:
         if active_loop is not None:
             return active_loop.results
 
+    def get_loops_state_dict(self) -> Dict[str, Any]:
+        return {
+            "fit_loop": self.fit_loop.state_dict(),
+            "validate_loop": self.validation_loop.state_dict(),
+            "test_loop": self.test_loop.state_dict(),
+        }
+
```
Comment (carmocca) [inline pytorch_lightning/trainer/properties.py:564]: I think this would fit better under the "State properties" section in this file

File: pytorch_lightning/trainer/properties.py [hunk 14]
```diff
@@ -555,6 +555,13 @@ def _results(self) -> Optional[ResultCollection]:
         if active_loop is not None:
             return active_loop.results
 
+    def get_loops_state_dict(self) -> Dict[str, Any]:
+        return {
+            "fit_loop": self.fit_loop.state_dict(),
+            "validate_loop": self.validation_loop.state_dict(),
```
Comment (carmocca) [inline pytorch_lightning/trainer/properties.py:561]: Consistency

```suggestion
            "validation_loop": self.validation_loop.state_dict(),
```

File: pytorch_lightning/loops/fit_loop.py [hunk 15]
```diff
@@ -274,3 +282,11 @@ def _check_checkpoint_callback(self, should_update: bool, is_last: bool = False)
 
             for cb in callbacks:
                 cb.on_validation_end(self.trainer, model)
+
+    def state_dict(self) -> Dict:
+        if not self.is_connected:
+            raise MisconfigurationException("The Trainer should be connected to loop to retrieve the state_dict.")
```
Comment (carmocca) [inline pytorch_lightning/loops/fit_loop.py:288]: Why is this check necessary?

File: pytorch_lightning/loops/base.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/loops/base.py:69]: Removed ! @awaelchli I will let you explore the corner case there.

Comment (tchaton) [reply pytorch_lightning/loops/base.py:69]: Done !

File: pytorch_lightning/loops/base.py (hunk 8 above)
Comment (tchaton) [reply pytorch_lightning/loops/base.py:52]: Removed !

File: pytorch_lightning/loops/batch/training_batch_loop.py (hunk 10 above)
Comment (tchaton) [reply pytorch_lightning/loops/batch/training_batch_loop.py:679]: Done !

File: pytorch_lightning/loops/batch/training_batch_loop.py (hunk 9 above)
Comment (tchaton) [reply pytorch_lightning/loops/batch/training_batch_loop.py:682]: Done !

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 11 above)
Comment (tchaton) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:430]: Done !

File: pytorch_lightning/loops/fit_loop.py (hunk 12 above)
Comment (tchaton) [reply pytorch_lightning/loops/fit_loop.py:106]: There is one setter for max_steps. I just added it for consistency.

File: pytorch_lightning/trainer/properties.py (hunk 13 above)
Comment (tchaton) [reply pytorch_lightning/trainer/properties.py:564]: Done !

File: pytorch_lightning/trainer/properties.py (hunk 14 above)
Comment (tchaton) [reply pytorch_lightning/trainer/properties.py:561]: Changed for `validate_loop` to match Trainer function names.

File: pytorch_lightning/loops/fit_loop.py (hunk 15 above)
Comment (tchaton) [reply pytorch_lightning/loops/fit_loop.py:288]: Removed !

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:51]: Done !

**Review by carmocca — APPROVED** [review]

File: tests/loops/test_loops.py [hunk 16]
```diff
@@ -0,0 +1,47 @@
+# Copyright The PyTorch Lightning team.
```
Comment (ananthsub) [inline tests/loops/test_loops.py:1]: nit: filename as `test_loop_state_dict` ?

**Review by ananthsub — COMMENTED** [review]
could you describe how the trainer will resume the states? at what point is the loop's `load_state_dict` called? 

also does what requirements does this impose on composition of loops?

File: pytorch_lightning/trainer/properties.py [hunk 17]
```diff
@@ -484,6 +484,13 @@ def sanity_checking(self, val: bool) -> None:
         elif self.sanity_checking:
             self.state.stage = None
 
+    def loops_state_dict(self) -> Dict[str, Any]:
+        return {
+            "fit_loop": self.fit_loop.state_dict(),
+            "validate_loop": self.validate_loop.state_dict(),
+            "test_loop": self.test_loop.state_dict(),
+        }
+
```
Comment (ananthsub) [inline pytorch_lightning/trainer/properties.py:493]: why does this need to be exposed as  a trainer property?

File: pytorch_lightning/loops/base.py [hunk 18]
```diff
@@ -59,6 +60,10 @@ def skip(self) -> bool:
     def connect(self, trainer: 'pl.Trainer', *args: Any, **kwargs: Any) -> None:
         """Connects Loop with all the necessary things like connectors and accelerators."""
         # TODO(@justusschock): Make the trainer a weakref/proxy
+        if not isinstance(trainer, pl.Trainer):
+            raise MisconfigurationException(
+                f"Loop {self.__class__.__name__} should be connected to a `Trainer`, found: {trainer}."
+            )
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/loops/base.py:66]: Yes, but it is slightly cleaner to enforce a strong check there.

File: tests/loops/test_loops.py (hunk 16 above)
Comment (tchaton) [reply tests/loops/test_loops.py:1]: Done !

File: pytorch_lightning/trainer/properties.py (hunk 17 above)
Comment (tchaton) [reply pytorch_lightning/trainer/properties.py:493]: Hey @ananthsub,

1. This isn't a strong requirements. In CheckpointConnector, we would have

```
def get_trainer_states(self):

    return {
        "loops": self.trainer.loops_state_dict(),
        ....
    }
```

2. I would say it is the target to save the `progress_tracking Dataclasses` + Results objects to enable full fail-tolerant training.

Comment (ananthsub) [reply pytorch_lightning/trainer/properties.py:493]: In that case I'd prefer not exposing this now, as it's only relevant for checkpointing. We could introduce this later if we absolutely need to

Comment (awaelchli) [reply pytorch_lightning/trainer/properties.py:493]: I would also prefer to  just let checkpoint connector directly access the loops to get the state dict, for now. 
We do the same for callbacks, precision, etc.

File: tests/loops/ test_loop_state_dict.py [hunk 19]
```diff
@@ -0,0 +1,47 @@
+# Copyright The PyTorch Lightning team.
```
Comment (awaelchli) [inline tests/loops/ test_loop_state_dict.py:1]: There is leading whitespace in the filename here, haha, how is this possible 😄

**Review by awaelchli — APPROVED** [review]

**Review by Borda — APPROVED** [review]

File: tests/loops/ test_loop_state_dict.py (hunk 19 above)
Comment (tchaton) [reply tests/loops/ test_loop_state_dict.py]: ahaha removed !

File: pytorch_lightning/trainer/properties.py (hunk 17 above)
Comment (carmocca) [reply pytorch_lightning/trainer/properties.py:493]:
```suggestion
```

File: pytorch_lightning/trainer/properties.py [hunk 20]
```diff
@@ -16,7 +16,7 @@
 from abc import ABC
 from argparse import ArgumentParser, Namespace
 from pathlib import Path
-from typing import cast, List, Optional, Type, TypeVar, Union
+from typing import Any, cast, Dict, List, Optional, Type, TypeVar, Union
```
Comment (carmocca) [inline pytorch_lightning/trainer/properties.py:19]:
```suggestion
from typing import cast, List, Optional, Type, TypeVar, Union
```
